code/game/objects/items/mail.dm

```diff
@@ -0,0 +1,282 @@
+	if(contents.len)
+		user.put_in_hands(contents[1])
+	playsound(loc, 'sound/items/poster_ripped.ogg', 50, TRUE)
+	qdel(src)
+
+/// Accepts a mob to initialize goodies for a piece of mail.
+/obj/item/mail/proc/initialize_for_recipient(mob/new_recipient)
+	recipient = new_recipient
+	name = "[initial(name)] for [new_recipient.real_name] ([new_recipient.job])"
+	var/list/goodies = generic_goodies
+
+	var/datum/job/this_job = SSjob.name_occupations[new_recipient.job]
+	if(this_job)
+		if(this_job.paycheck_department && department_colors[this_job.paycheck_department])
+			color = department_colors[this_job.paycheck_department]
+		var/list/job_goodies = this_job.get_mail_goodies()
+		if(LAZYLEN(job_goodies))
+			// certain roles and jobs (prisoner) do not receive generic gifts.
+			if(this_job.exclusive_mail_goodies)
+				goodies = job_goodies
+			else
+				goodies += job_goodies
+
+	for(var/iterator = 0, iterator < goodie_count, iterator++)
+		var/target_good = pickweight(goodies)
+		if(ispath(target_good, /datum/reagent))
+			var/obj/item/reagent_containers/target_container = new /obj/item/reagent_containers/glass/bottle(src)
+			target_container.reagents.add_reagent(target_good, target_container.volume)
+			target_container.name = "[target_container.reagents.reagent_list[1].name] bottle"
+			new_recipient.log_message("[key_name(new_recipient)] received reagent container [target_container.name] in the mail ([target_good])", LOG_GAME)
+		else
+			var/atom/movable/target_atom = new target_good(src)
+			new_recipient.log_message("[key_name(new_recipient)] received [target_atom.name] in the mail ([target_good])", LOG_GAME)
+
+	return TRUE
+
+/// Alternate setup, just complete garbage inside and anyone can open
+/obj/item/mail/proc/junk_mail()
+
+	var/obj/junk = /obj/item/paper/fluff/junkmail_generic
+	var/special_name = FALSE
+
+	if(prob(25))
+		special_name = TRUE
+		junk = pick(list(/obj/item/paper/pamphlet/gateway, /obj/item/paper/pamphlet/violent_video_games, /obj/item/paper/fluff/junkmail_redpill, /obj/effect/decal/cleanable/ash))
+
+	var/list/junk_names = list(
+		/obj/item/paper/pamphlet/gateway = "[initial(name)] for [pick(GLOB.adjectives)] adventurers",
+		/obj/item/paper/pamphlet/violent_video_games = "[initial(name)] for the truth about the arcade centcom doesn't want to hear",
+		/obj/item/paper/fluff/junkmail_redpill = "[initial(name)] for those feeling [pick(GLOB.adjectives)] working at Nanotrasen",
+		/obj/effect/decal/cleanable/ash = "[initial(name)] with INCREDIBLY IMPORTANT ARTIFACT- DELIVER TO SCIENCE DIVISION. HANDLE WITH CARE.",
+	)
+
+	color = pick(department_colors) //eh, who gives a shit.
+	name = special_name ? junk_names[junk] : "important [initial(name)]"
+
+	junk = new junk(src)
+	return TRUE
+
+/obj/item/mail/proc/disposal_handling(obj/structure/disposalholder/disposal_holder, obj/machinery/disposal/disposal_machine, hasmob)
+	SIGNAL_HANDLER
+	if(!hasmob)
+		disposal_holder.destinationTag = sort_tag
```

Comment:
Something about this is failing to sort letters and mail. Thoughts, @Mothblocks ?

Reply:
Not sure! Your best bet is to just run a debugger or throw a bunch of logs to see if it runs, and what the values are by the time of sorting.